### PR TITLE
chore: migrate src/services

### PR DESCRIPTION
**Pull request by flochtililoch**

### CODE REVIEW

File: src/services/navigator/helpers.test.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import * as DomErrors from 'hyperview/src/services/dom/errors';
 import * as Errors from './errors';
 import * as Namespaces from '../namespaces';
 import * as Types from './types';
-import * as TypesLegacy from '../../types-legacy';
+import * as TypesLegacy from '../../types';
```
Comment (hgray-instawork) [inline src/services/navigator/helpers.test.ts:5]: Can we correct these paths now that the tsconfig is updated?

**Review by hgray-instawork — APPROVED** [review]

File: src/services/behaviors/index.ts [hunk 2]
```diff
@@ -28,15 +20,18 @@ export const toggleIndicators = (
   ids: Array<string>,
   showIndicators: boolean,
   root: Document,
-): Document =>
-  ids.reduce((newRoot, id) => {
-    const indicatorElement: ?Element = newRoot.getElementById(id);
+): Document => {
+  return ids.reduce((newRoot, id) => {
```
Comment (flochtililoch) [inline src/services/behaviors/index.ts:24]: revert

File: src/services/behaviors/index.ts [hunk 3]
```diff
@@ -81,15 +76,16 @@ export const performUpdate = (
     const { parentNode } = targetElement;
     if (parentNode) {
       parentNode.replaceChild(newElement, targetElement);
-      return shallowCloneToRoot((parentNode: any));
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      return shallowCloneToRoot(parentNode as any);
```
Comment (flochtililoch) [inline src/services/behaviors/index.ts:80]: anyway we can cast as `Document`?

File: src/services/behaviors/index.ts [hunk 4]
```diff
@@ -133,14 +129,16 @@ export const trigger = (
...
-    onUpdate(href, action, element, {
-      behaviorElement,
-      delay,
-      hideIndicatorIds,
-      once,
-      showIndicatorIds,
-      targetId,
-      verb,
-    });
+    if (onUpdate) {
```
Comment (flochtililoch) [inline src/services/behaviors/index.ts:132]: can we make it non optional?

File: src/services/components/index.ts [hunk 5]
```diff
@@ -52,12 +50,23 @@ const reducer = (registry: ComponentRegistry, component: HvComponent) => ({
 
 export const getRegistry = (
   components: HvComponent[] = [],
-): ComponentRegistry =>
-  [...HYPERVIEW_COMPONENTS, ...components].reduce(reducer, {});
+): ComponentRegistry => {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return [...HYPERVIEW_COMPONENTS, ...components].reduce<Record<string, any>>(
```
Comment (flochtililoch) [inline src/services/components/index.ts:55]: revert

File: src/services/components/index.ts [hunk 6]
```diff
@@ -52,12 +50,23 @@ const reducer = (registry: ComponentRegistry, component: HvComponent) => ({
...
+};
 
 export const getFormRegistry = (
   components: HvComponent[] = [],
-): ComponentRegistry =>
-  [...HYPERVIEW_COMPONENTS, ...components]
-    .filter(c => Object.prototype.hasOwnProperty.call(c, 'getFormInputValues'))
-    .reduce(reducer, {});
+): ComponentRegistry => {
+  return (
```
Comment (flochtililoch) [inline src/services/components/index.ts:64]: revert

File: src/services/components/internal.ts [hunk 7]
```diff
@@ -12,11 +10,17 @@ import type { HvComponent } from 'hyperview/src/types';
 
 export const registerComponent = (
   component: HvComponent,
-): { [string]: HvComponent } =>
-  [component.localName, ...(component.localNameAliases || [])].reduce(
+): {
+  [key: string]: HvComponent;
+} => {
+  return [component.localName, ...(component.localNameAliases || [])].reduce<
```
Comment (flochtililoch) [inline src/services/components/internal.ts:16]: revert

File: src/services/dom/errors.ts [hunk 8]
```diff
@@ -8,6 +6,7 @@
  *
  */
 
+// eslint-disable-next-line max-classes-per-file
```
Comment (flochtililoch) [inline src/services/dom/errors.ts:9]: I think this only throws a warning, not an error - can be removed

File: src/services/dom/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+import * as Namespaces from 'hyperview/src/services/namespaces';
+import type { LocalName, NamespaceURI, NodeType } from 'hyperview/src/types';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const getBehaviorElements = (element: any) => {
```
Comment (flochtililoch) [inline src/services/dom/helpers.ts:13]: let's explore the diff in more details to understand why it's showing as added.

File: src/services/dom/index.test.ts [hunk 10]
```diff
@@ -14,19 +12,20 @@ import { X_RESPONSE_STALE_REASON } from './types';
...
-const mockExpectedDocument = { foo: 'bar' };
+const mockExpectedDocument = { foo: 'bar' } as const;
 const mockParseFromString = jest.fn().mockReturnValue(mockExpectedDocument);
 jest.mock('@instawork/xmldom', () => {
   const DOMParser = () => null;
-  DOMParser.prototype.parseFromString = (...args) =>
-    mockParseFromString.apply(this, args);
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  DOMParser.prototype.parseFromString = (...args: any) => {
+    return mockParseFromString.apply(this, args);
```
Comment (flochtililoch) [inline src/services/dom/index.test.ts:21]: revert

File: src/services/dom/parser.ts [hunk 11]
```diff
@@ -49,14 +46,14 @@ const parser = new DOMParser({
...
 
-  onAfterParse: ?BeforeAfterParseHandler;
+  onAfterParse: BeforeAfterParseHandler | null | undefined;
 
   constructor(
     fetch: Fetch,
-    onBeforeParse: ?BeforeAfterParseHandler,
-    onAfterParse: ?BeforeAfterParseHandler,
+    onBeforeParse?: BeforeAfterParseHandler | null,
+    onAfterParse?: BeforeAfterParseHandler | null,
```
Comment (flochtililoch) [inline src/services/dom/parser.ts:56]: signature changed

File: src/services/dom/parser.ts [hunk 12]
```diff
@@ -134,9 +138,12 @@ export class Parser {
 
     const screenElement = getFirstTag(docElement, LOCAL_NAME.SCREEN);
     const navigatorElement = getFirstTag(docElement, LOCAL_NAME.NAVIGATOR);
-    if (!screenElement && !navigatorElement) {
+    if (!screenElement) {
+      throw new Errors.XMLRequiredElementNotFound(LOCAL_NAME.SCREEN, baseUrl);
+    }
+    if (!navigatorElement) {
```
Comment (flochtililoch) [inline src/services/dom/parser.ts:144]: is this refactor needed?

File: src/services/dom/parser.ts [hunk 13]
```diff
@@ -155,19 +162,28 @@ export class Parser {
         );
       }
     } else {
-      throw new Errors.XMLRequiredElementNotFound(
-        `${LOCAL_NAME.SCREEN}/${LOCAL_NAME.NAVIGATOR}`,
-        baseUrl,
-      );
+      if (!screenElement) {
```
Comment (flochtililoch) [inline src/services/dom/parser.ts:165]: same question here

File: src/services/events/index.ts [hunk 14]
```diff
@@ -20,8 +18,10 @@ export const dispatch = (eventName: string) => {
   tinyEmitter.emit(ON_EVENT_DISPATCH, eventName);
 };
 
-export const subscribe = (callback: (eventName: string) => void) =>
-  tinyEmitter.on(ON_EVENT_DISPATCH, callback);
+export const subscribe = (callback: (eventName: string) => void) => {
+  return tinyEmitter.on(ON_EVENT_DISPATCH, callback);
```
Comment (flochtililoch) [inline src/services/events/index.ts:22]: revert

File: src/services/events/index.ts [hunk 15]
```diff
@@ -20,8 +18,10 @@ export const dispatch = (eventName: string) => {
...
+};
 
-export const unsubscribe = (callback: (eventName: string) => void) =>
-  tinyEmitter.off(ON_EVENT_DISPATCH, callback);
+export const unsubscribe = (callback: (eventName: string) => void) => {
+  return tinyEmitter.off(ON_EVENT_DISPATCH, callback);
```
Comment (flochtililoch) [inline src/services/events/index.ts:26]: revert

File: src/services/index.ts [hunk 16]
```diff
@@ -127,23 +126,25 @@ export const createProps = (
...
   props.style = createStyleProp(element, stylesheets, options);
   const testProps = createTestProps(element);
-  // $FlowFixMe
   return { ...props, ...testProps };
 };
 
-export const later = (delayMs: number): Promise<void> =>
-  new Promise(resolve => setTimeout(resolve, delayMs));
+export const later = (delayMs: number): Promise<void> => {
+  return new Promise((resolve: (result?: Promise<never>) => void) => {
```
Comment (flochtililoch) [inline src/services/index.ts:133]: revert

File: src/services/index.ts [hunk 17]
```diff
@@ -325,10 +329,11 @@ export const getNameValueFormInputValues = (
   return [];
 };
 
-export const encodeXml = (xml: string): string =>
-  xml
+export const encodeXml = (xml: string): string => {
+  return xml
```
Comment (flochtililoch) [inline src/services/index.ts:333]: revert

File: src/services/inline-context/helpers.ts [hunk 18]
```diff
@@ -15,14 +13,16 @@ export const SPACE = ' ';
 /**
  * Removes any space that follows a line break
  */
-export const ignoreSpacesAfterLineBreak = (input: string): string =>
-  input.replace(/\n\s+/g, LINE_BREAK);
+export const ignoreSpacesAfterLineBreak = (input: string): string => {
+  return input.replace(/\n\s+/g, LINE_BREAK);
```
Comment (flochtililoch) [inline src/services/inline-context/helpers.ts:17]: revert

File: src/services/inline-context/helpers.ts [hunk 19]
```diff
@@ -15,14 +13,16 @@ export const SPACE = ' ';
...
+};
 
 /**
  * Replaces any line break by a space
  */
-export const convertLineBreaksIntoSpaces = (input: string): string =>
-  input.replace(/\n/g, SPACE);
+export const convertLineBreaksIntoSpaces = (input: string): string => {
+  return input.replace(/\n/g, SPACE);
```
Comment (flochtililoch) [inline src/services/inline-context/helpers.ts:24]: revert

File: src/services/inline-context/index.ts [hunk 20]
```diff
@@ -74,15 +72,15 @@ import { NODE_TYPE } from 'hyperview/src/types';
...
-  const nodes = Dom.preorder(root, NODE_TYPE.TEXT_NODE);
+export const formatter = (root: Element | Document): [Node[], string[]] => {
+  const nodes = preorder(root, NODE_TYPE.TEXT_NODE);
 
-  const nodeValues: string[] = Array.from(nodes)
-    .map((node: Node): string =>
-      node && node.nodeValue
+  const nodeValues: string[] = Array.from<Node>(nodes)
+    .map((node: Node): string => {
+      return node && node.nodeValue
```
Comment (flochtililoch) [inline src/services/inline-context/index.ts:80]: revert

File: src/services/navigation/index.ts [hunk 21]
```diff
@@ -42,7 +41,9 @@ export default class Navigation {
     this.document = document;
   };
 
-  getPreloadScreen = (id: number): ?Element => preloadScreens[id];
+  getPreloadScreen = (id: number): Element | null | undefined => {
+    return preloadScreens[id];
```
Comment (flochtililoch) [inline src/services/navigation/index.ts:45]: revert

File: src/services/navigation/index.ts [hunk 22]
```diff
@@ -52,7 +53,9 @@ export default class Navigation {
     delete preloadScreens[id];
   };
 
-  getRouteKey = (href: string): ?string => routeKeys[getHrefKey(href)];
+  getRouteKey = (href: string): string | null | undefined => {
+    return routeKeys[getHrefKey(href)];
```
Comment (flochtililoch) [inline src/services/navigation/index.ts:57]: revert

File: src/services/navigation/index.ts [hunk 23]
```diff
@@ -82,13 +88,13 @@ export default class Navigation {
...
     if (showIndicatorId && this.document) {
-      const screens: NodeList<Element> = this.document.getElementsByTagNameNS(
+      const screens: HTMLCollectionOf<Element> = this.document.getElementsByTagNameNS(
         Namespaces.HYPERVIEW,
         'screen',
       );
-      const loadingScreen: ?Element = Array.from(screens).find(
-        s => s && s.getAttribute('id') === showIndicatorId,
-      );
+      const loadingScreen: Element | null | undefined = Array.from(
```
Comment (flochtililoch) [inline src/services/navigation/index.ts:95]: needed?

File: src/services/navigation/index.ts [hunk 24]
```diff
@@ -82,13 +88,13 @@ export default class Navigation {
 
     let preloadScreen = null;
     if (showIndicatorId && this.document) {
-      const screens: NodeList<Element> = this.document.getElementsByTagNameNS(
+      const screens: HTMLCollectionOf<Element> = this.document.getElementsByTagNameNS(
```
Comment (flochtililoch) [inline src/services/navigation/index.ts:91]: can `NodeListOf` be used instead?

File: src/services/navigator/helpers.test.ts (hunk 1 above)
Comment (flochtililoch) [reply src/services/navigator/helpers.test.ts:5]: let's address this separately after the migration is merged on master

File: src/services/xml/index.ts [hunk 25]
```diff
@@ -12,5 +10,6 @@
  * Splits a string by any whitespace and returns an array
  * containing only the non-whitespace tokens.
  */
-export const splitAttributeList = (attr: string): Array<string> =>
-  attr.match(/\S+/g) || [];
+export const splitAttributeList = (attr: string): Array<string> => {
+  return attr.match(/\S+/g) || [];
```
Comment (flochtililoch) [inline src/services/xml/index.ts:14]: revert

File: src/services/behaviors/index.ts (hunk 3 above)
Comment (hgray-instawork) [reply src/services/behaviors/index.ts:80]: Probably change to
`return shallowCloneToRoot(parentNode) as Document;`

File: src/services/dom/errors.ts (hunk 8 above)
Comment (hgray-instawork) [reply src/services/dom/errors.ts:9]: It makes the file very tough to read. The entire thing is underlined in yellow

File: src/services/navigation/index.ts (hunk 24 above)
Comment (flochtililoch) [reply src/services/navigation/index.ts:91]: no

**Pull request merged**